solana/pyth2wormhole/client/src/batch_state.rs

```diff
@@ -40,63 +43,46 @@ impl<'a> BatchState<'a> {
-        self.status_changed_at = Instant::now();
-        self.status = s;
-    }
 
     /// Evaluate the configured attestation conditions for this
     /// batch. RPC is used to update last known state. Returns
     /// Some("<reason>") if any trigger condition was met. Only the
     /// first encountered condition is mentioned.
-    pub fn should_resend(&mut self, c: &RpcClient) -> Option<String> {
+    pub async fn should_resend(&mut self, c: &RpcClient) -> Option<String> {
         let mut ret = None;
 
         let sym_count = self.symbols.len();
-        let mut new_symbol_states: Vec<Option<PriceAccount>> = Vec::with_capacity(sym_count);
-        for (idx, sym) in self.symbols.iter().enumerate() {
-            let new_state = match c
-                .get_account_data(&sym.price_addr)
-                .map_err(|e| e.to_string())
-                .and_then(|bytes| {
-                    pyth_sdk_solana::state::load_price_account(&bytes)
-                        .map(|state| state.clone())
-                        .map_err(|e| e.to_string())
-                }) {
-                Ok(state) => Some(state),
-                Err(e) => {
-                    warn!(
-                        "Symbol {} ({}/{}): Could not look up state: {}",
-                        sym.name
-                            .as_ref()
-                            .unwrap_or(&format!("Unnamed product {}", sym.product_addr)),
-                        idx + 1,
-                        sym_count,
-                        e.to_string()
-                    );
-                    None
-                }
-            };
-
-            new_symbol_states.push(new_state);
+        let mut new_symbol_states: Vec<Option<PriceAccount>> = vec![None; sym_count];
+        let pubkeys: Vec<_> = self.symbols.iter().map(|s| s.price_addr).collect();
+
+        // Always learn the current on-chain state for each symbol
+        match c.get_multiple_accounts(&pubkeys).await {
+            Ok(acc_opts) => {
+                new_symbol_states = acc_opts
+                    .into_iter()
+                    .enumerate()
+                    .map(|(idx, opt)| {
+                        // Take each Some(acc), make it None and log on load_price_account() error
+                        opt.and_then(|acc| {
+                            pyth_sdk_solana::state::load_price_account(&acc.data)
+                                .cloned() // load_price_account() transmutes the data reference into another reference, and owning acc_opts is not enough
+                                .map_err(|e| {
+                                    warn!("Could not parse symbol {}/{}: {}", idx, sym_count, e);
+                                    e
+                                })
+                                .ok() // Err becomes None
+                        })
+                    })
+                    .collect();
+            }
+            Err(e) => warn!("Could not look up any symbols on-chain: {}", e),
```

Comment:
You could move this below the "min interval" check, which will potentially reduce the number of RPC calls.

Reply:
Sorry if I'm misunderstanding, but I think all this loop is doing is updating `new_symbol_states`. We don't reference `new_symbol_states` unless the "min interval" check below passes, so I'm not sure what we gain by updating it every time? What "other conditions" are you referring to? Sorry if I'm missing something :)

Reply:
The general idea is to populate `ret` instead of returning and let it be returned at the very end. `new_symbol_states`  is referenced when we check that a condition was found. We use it update the state only if a condition is met. This way we can evaluate conditions since last attestation as opposed to since last lookup. The other conditions are publish_time change and price value change

Reply:
i.e. we have three possible conditions that can trigger an attestation:
1. min_interval elapsed since last attestation
2. publish_time changed by at least the specified value
3. price value changed by at least the specified percentage

Reply:
I see, thank-you!